subsys/caf/modules/sensor_sampler.c

```diff
@@ -61,26 +80,142 @@ static void send_sensor_event(const char *descr, struct sensor_value *data, size
+}
+
+static size_t get_sensor_data_cnt(const struct sensor_config *sc)
 {
 	size_t data_cnt = 0;
 
 	for (size_t i = 0; i < sc->chan_cnt; i++) {
 		data_cnt += sc->chans[i].data_cnt;
 	}
 
+	return data_cnt;
+}
+
+static bool is_active(const struct sensor_config *sc, double curr, double prev)
+{
+	enum trigger_activation_type type = sc->trigger->activation.type;
+	double tresh = sc->trigger->activation.tresh;
+
+	switch (type) {
+	case PERCENTAGE_DEVIATION:
+		return FABS(curr - prev) < FABS(prev * tresh/100.0);
+	case ABSOLUTE_DEVIATION:
+		return FABS(curr - prev) < FABS(tresh);
+	default:
+		report_error();
+	}
+
+	return false;
+}
+
+static bool can_sensor_sleep(const struct sensor_config *sc,
+			     struct sensor_data *sd,
+			     struct sensor_value *new)
+{
+	size_t data_cnt = get_sensor_data_cnt(sc);
+	bool sleep = true;
+
+	for (size_t i = 0; i < data_cnt; i++) {
+		double prev = sensor_value_to_double(&sd->prev[i]);
+		double curr = sensor_value_to_double(&new[i]);
+
+		if (!is_active(sc, curr, prev)) {
+			sleep = false;
+			break;
+		}
+	}
+
+	if (sleep) {
+		unsigned int timeout = sc->trigger->activation.timeout_ms;
+		unsigned int period = sc->sampling_period_ms;
+
+		sd->sleep_cnt++;
+		if (sd->sleep_cnt >= timeout/period) {
+			sd->sleep_cnt = 0;
+			return true;
+		}
+	} else {
+		memcpy(sd->prev, new, data_cnt*sizeof(struct sensor_value));
+		sd->sleep_cnt = 0;
+	}
+
+	return false;
+}
+
+static void trigger_handler(const struct device *dev, struct sensor_trigger *const trigger)
+{
+	int err;
+
+	err = sensor_trigger_set(dev, trigger, NULL);
+
+	struct sensor_data *sd = get_sensor_data(dev);
+
+	sd->sample_timeout = k_uptime_get();
+
+	err = !atomic_cas(&sd->state, SENSOR_SLEEP, SENSOR_ACTIVE);
+	if (err) {
+		__ASSERT_NO_MSG(!err);
+		report_error();
+		return;
+	}
+
+	LOG_INF("%s sensor woken up", dev->name);
+	k_sem_give(&sample);
+}
+
+static int sample_sensor(struct sensor_data *sd, const struct sensor_config *sc)
+{
 	int err = 0;
 	size_t data_idx = 0;
+	size_t data_cnt = get_sensor_data_cnt(sc);
 	struct sensor_value data[data_cnt];
 
-	err = sensor_sample_fetch(dev);
+	err = sensor_sample_fetch(sd->dev);
 	for (size_t i = 0; !err && (i < sc->chan_cnt); i++) {
 		const struct sampled_channel *sampled_chan = &sc->chans[i];
 
-		err = sensor_channel_get(dev, sampled_chan->chan, &data[data_idx]);
+		err = sensor_channel_get(sd->dev, sampled_chan->chan, &data[data_idx]);
 		data_idx += sampled_chan->data_cnt;
 	}
 
+	if (sc->trigger) {
+		if (can_sensor_sleep(sc, sd, data)) {
+			atomic_set(&sd->state, SENSOR_SLEEP);
+			err = sensor_trigger_set(sd->dev, &sc->trigger->cfg, trigger_handler);
+			if (err) {
+				if (err == -ENOTSUP) {
+					LOG_ERR("Check if trigger is enabled for %s",
+						sc->dev_name);
+				}
+				__ASSERT_NO_MSG(!err);
+				atomic_set(&sd->state, SENSOR_ERROR);
+				report_error();
+				return err;
+			}
+			LOG_INF("%s sensor put to sleep", sc->dev_name);
+		} else {
+			k_sem_give(&sample);
```

Comment:
I review bottom - up so my comment is after I have seen other places.
I am greatly against this way of tiggering the semaphore. It seems needlessly complicated.

In short - main loop should sleep until next sampling time, unless sensor that was sleeping is triggered in which case the main loop should be instantly woken up to check what requires sampling.

In other words what you need is to create a function that will calculate the semaphore timeout (the lowest of the times for next sample for all sensors or K_FOREVER is all sensors sleep). All triggers should update the respective sensor state (request sample at current time) and ping the main loop (give semaphore).

The outcome will be sampling done from one context and much simpler state management.